### ISSUE TITLE
Questions about the moving average

### ISSUE THREAD

**Issue by Chloe1997** [qinenergy/cotta#9]
Hi! Thanks for sharing your code. This work is interesting! I have some questions about the updating of the teacher model. It would be great if you could answer them.
The original Mean Teacher is trained with both labeled and unlabeled data. The consistency loss then calculates the difference between the outputs from the student and the teacher model. The author did not have high confidence about the feasibility of training Mean teacher without any labeled data. [https://github.com/CuriousAI/mean-teacher/issues/24](https://github.com/CuriousAI/mean-teacher/issues/24)  
In your code, the student model update with the softmax_entropy and the teacher model accumulated the historical weights from the student model. I am wonder how could you avoid such performance degradation on the teacher model when there is no other guidance on the student model.
I have tried to implement the Mean Teacher model without labeled data training on the student model. The training results shows that the performance of the teacher model dropped when updated with exponential moving average every steps. Had you ever encountered such problems?

**Comment by qinenergy** [comment]
Hi, thanks for the question.

The mean teacher paper focuses on the semi-supervised learning setup. This work considers the test-time domain adaptation setup. The main difference is: 
+ Labeled data is used in our case to pre-train the source model, but the labeled data cannot be accessed during test time for adaptation. 
+ The domain shift is there in our case between the labeled(source) and unlabeled(target) data. 

Therefore the starting point of these two setups are quite different. 
+ In our case, the source pre-trained model is **already well-trained for the labeled data (source domain)**, and can be considered as a good local optima for the source data. The aim of test-time DA is to start from this point, and find a local optima for the current target domain by using the unlabeled target data. Since the target domain is closely related to the source domain, we can assume that a desired solution shall be nearby. To avoid forgetting the knowledge from the source labeled data, we also deploy the stochastic restoration to the source model. 
+ In semi-supervised learning, models are often **randomly** initialized (or using some completely unrelated general weights), and the aim is to find a good solution for one domain (which is the same for both labeled and unlabeled) by making use of both the labeled and unlabeled data. If you don't use the labeled data in this case in your mentioned experiments, then it is apparently much harder to optimize because you don't have enough supervision and it is no where near the desired minima. Errors can then easily be accumulated, student performance may drop and so comes the mean teacher performance. 

To sum up, we make use of the labeled data by using the source pre-trained model which is trained by the source labeled data. 
+ The source pre-trained model is important because it provides a good starting point which makes the mean teacher optimization much easier. 
+ It is also important as it provides the knowledge and supervision on the classes. 
+ This is also why we add the stochastic restoration component to further ensure that the knowledge from the labeled data is not forgotten too much. 

As shown in the ablation in Table 2, mean teacher is quite useful for our setup.

**Comment by Chloe1997** [comment]
Thanks for the detail explanations! 
There are some questions I would like to ask:
- The well pre-trained source model was used for the my experiments motioned above.  Unfortunately, the teacher model could not have stable training as shown in the figure.
![image](https://user-images.githubusercontent.com/42937407/176607582-27acc9dd-b670-4c3d-b3e8-9183b51b4b84.png)
The accuracy of pseudo labels provided by the teacher model are around 0.93%/ batch after filtering with confidence threshold 0.8. I am wonder that the inaccurate pseudo labels affect the performance of the teacher model. 
In your work, the correctness of the pseudo labels is critical and the proposed **Weight-Averaged** and  **Augmentation-Averaged** improve the quality of the pseudo labels. Could you please provide the average accuracy of the pseudo labels during online adaptation? 

- **Exponential Moving Average Normalization for Self-supervised and Semi-supervised Learning** updates the teacher model not only with the parameters but also Batch normalization statistics. Do you think updating with BN statics could further improve the mean teacher framework?

- Also, I would like to know did you ever try the memory aware synapse to prevent the catastrophic forgetting.  I think it would be great to compare the effectivity of Stochastic Restoration and other lifelong learning methods.

Thank your for the replies! This is a really nice work.

**Comment by qinenergy** [comment]
Regarding your questions for your own setups/experiments:

+ I don't think the average accuracy of the pseudo labels is really comparable across different tasks. Based on your accuracy plots for your own experiments, I would say that you should look at the plots without smoothing. It seems that the accuracy is varying too much across different batches/steps, you may want to take a look at your batchsize and data sampling process.
+ It really depends on your task. In our work, we are always using the batch BN statics from the current batch because the distribution is continually shifting. BN is indeed an important factor if you have some distribution shift. You will need to find the best solution to your own problem.  
+ No, we did not try the memory aware synapse method, but we did have a try on LwF(learning without forgetting). You can find it in the supplementary. 

Please let me know if you have questions directly related to the paper.

**Comment by Chloe1997** [comment]
Thanks for the quick replies and advise! Great work and I am really excited to see more works from your team!
Thank you!
